src/ast/analysis.rs

```diff
@@ -0,0 +1,100 @@
+//! Diverse checks that need to be done before lowering.
+use crate::{ast::{Located, Block, Node, Name, visitor::NodeVisitor}, ty::{Binding,Ty}};
+
+
+/// Checks if a function is recursive or not.
```

Comment:
@DarkDrek could you give this a look? I'm worried that it might not work as intended and ends in some sort of ICE.

Reply:
I looked at it (`RecursionChecker` and `NodeVisitor`) and gave a few comments why do you think it will give an ICE?

Reply:
This is just some gut feeling so take it with a grain of salt.

If for some weird reason `RecursionChecker` ends up saying that a function is not recursive an the function is. That means that it won't get wrapped in a `Term::Fix`. Which also means that the place where the function calls itself will have an unbounded name. 

This could cause the type checker to panic when it's checking the body of the recursive function, or in the best case scenario to actually say that the name of the function is unbounded.

But in a total disaster scenario where the type checker doesn't realize that the name is unbounded. Then the lowering to the `LIR` will panic, or worse it will try to execute the program and panic because there aren't enough indices for all the variables.

I don't have any hard reasons to believe this will fail. It is only that the notion of scoping inside the `RecursionChecker` has to coincide with the notion of scoping in the lower representations.